subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java

```diff
@@ -190,4 +219,30 @@ public void validate(String propertyName, boolean optional, ValidationAction val
             }
         }
     }
+
+    private static class ImplementationPropertyValue implements ValidatingValue {
+
+        private final HashCode value;
+
+        public ImplementationPropertyValue(Object bean, ClassLoaderHierarchyHasher classLoaderHierarchyHasher) {
+            BuildCacheHasher buildCacheHasher = new DefaultBuildCacheHasher();
+            HashCode classLoaderHash = classLoaderHierarchyHasher.getClassLoaderHash(bean.getClass().getClassLoader());
+            String typeName = bean.getClass().getName();
+            if (classLoaderHash != null) {
+                new ImplementationSnapshot(typeName, classLoaderHash).appendToHasher(buildCacheHasher);
+            } else {
+                buildCacheHasher.putString(typeName);
+            }
+            this.value = buildCacheHasher.hash();
+        }
```

Comment:
We could. I think calculating this hash is pretty cheap. It is hashing a String (the classname) and looking up and hashing a hash code. We also don't cache this in [`CacheBackedTaskHistoryRepository.createExecution`](https://github.com/gradle/gradle/blob/af9c80a3d7308019234ce097a21d52c42f1fdc2c/subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepository.java#L137).
We could introduce a `ClassImplementationHashCache` and use that one going forward.
WDYT? Should we introduce the cache as part of this change?

Reply:
It would be nice to measure first, but we actually create quite a bit of stuff around a new `DefaultBuildCacheHasher` that would be nice to avoid. I think this location is much more important to cache than `CacheBackedTaskHistoryRepository.createExecution`, because that guy gets called once for every execute task, while this one gets called for every nested property value, could end up being a lot more (especially if we start using rich inputs more).

Reply:
I created a `ClassImplemenationHasher` which caches the result of hashing an implemenation.